File: sphere.cpp
#include <cmath>

#include "sphere.hpp"

Sphere::Sphere(const Point &center, double radius, const Vector &color)
    : center(center), radius(radius), color(color) {}

bool Sphere::intersect(const Point &rayOrigin, const Vector &rayDirection, double &t) const
{
    // Calcula o vetor entre o centro da esfera e a origem do raio
    Vector rayToCenter = center.difference(rayOrigin);

    // Calcula o coeficiente 'a' da equação quadrática
    double a = rayDirection.dot(rayDirection);

    // Calcula o coeficiente 'b' da equação quadrática
    double b = 2.0 * rayToCenter.dot(rayDirection);

    // Calcula o coeficiente 'c' da equação quadrática
    double c = rayToCenter.dot(rayToCenter) - radius * radius;

    // Calcula o discriminante da equação quadrática
    double discriminant = b * b - 4 * a * c;

    // Verifica se há interseção
    if (discriminant > 0)
    {
        // Calcula as soluções da equação quadrática
        double t1 = (-b - std::sqrt(discriminant)) / (2.0 * a);
        double t2 = (-b + std::sqrt(discriminant)) / (2.0 * a);

        // Verifica se as soluções estão na frente do observador e escolhe a menor
        if (t1 > 0.0 || t2 > 0.0)
        {
            t = (t1 < t2) ? t1 : t2;
            return true;
        }
    }

    return false; // Não houve interseção
}

File: src/sphere.py
import math

from src.point import Point
from src.shape import Shape
from src.vector import Vector


class Sphere(Shape):
    def __init__(self, center: Point, radius: float, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.center = center
        self.radius = radius

    def intersect(self, ray_origin: Point, ray_direction: Vector) -> tuple[bool, float]:
        ray_to_center: Vector = self.center.difference(ray_origin)

        a = ray_direction.dot(ray_direction)
        b = -2 * ray_to_center.dot(ray_direction)
        c = ray_to_center.dot(ray_to_center) - self.radius * self.radius

        discriminant = b * b - 4 * a * c

        if discriminant > 0:
            discriminant_sqrt = math.sqrt(discriminant)
            t1 = (-b - discriminant_sqrt) / (2 * a)
            t2 = (-b + discriminant_sqrt) / (2 * a)
            if t1 > 0 or t2 > 0:
                return True, min(t1, t2)
        return False, 0
